parsers/events.js

```diff
@@ -0,0 +1,106 @@
+'use strict';
+
+let addedCharacters = 0;
+
+function splitPreviousEvent(previousEvent, voicesAddedToEvent, eventIndex) {
+  const voicesToAdd = [];
+
+  let hasAddedCharacter = false;
+  // Split any previous events
+  previousEvent.forEach((voice) => {
+    if (voicesAddedToEvent.indexOf(voice.voice) === -1 &&
+        eventIndex < voice.index + voice.content.toString().length) {
+      // Split voice from previous event, add remainder to this event.
+      const splitIndex = eventIndex - voice.index + addedCharacters;
+      const event = {
+        index: eventIndex + addedCharacters,
+        voice: voice.voice,
+        content: `-${voice.content.substring(splitIndex)}`
+      };
+      voice.content = voice.content.substring(0, splitIndex);
+
+      voicesToAdd.push(event);
+
+      // Update index position to account for added character.
+      if (!hasAddedCharacter) {
+        addedCharacters++;
+        hasAddedCharacter = true;
+      }
+    }
+  });
+
+  return voicesToAdd;
+}
+
+function addEvents(verse, longestEvent) {
+  const voices = [];
+  const voicesAdded = [];
+
+  verse.forEach((events, voiceName) => {
+    if (events.length === 0) {
+      return;
+    }
+
+    const event = events[0];
+    // Only include an event if it falls between the longest event's start and end index.
+    if (longestEvent.index <= event.index &&
+        event.index < longestEvent.index + longestEvent.content.toString().length) {
+      const eventToAdd = events.shift();
+      eventToAdd.voice = voiceName;
+      eventToAdd.index += addedCharacters;
+      voices.push(eventToAdd);
+      voicesAdded.push(voiceName);
+    }
+  });
+
+  return { voicesAdded, voices };
+}
+
+function createEventFromVerse(verse, voiceOrder, previousEvent = []) {
+  const firstEventOfEachVoice = Array.from(verse.values()).map((voiceArr) => voiceArr[0]);
+
+  // Find the minimum index of all voices, since we want to parse events in the order they happen.
+  const eventIndex = firstEventOfEachVoice.reduce((acc, voice) => {
```

Comment:
When I was first imagining the implementation of this logic, I had the intuition that you'd want your events in sorted order. That was why I had merged all the voices in the phrase and sorted them.

Question: what would this algorithm look like if you started with that data structure instead?